src/main/java/org/springframework/retry/support/MetricsRetryListener.java

```diff
@@ -0,0 +1,132 @@
+ * call means no counts</li>
+ * <li>{@code exception} - the thrown back to the caller (after all the retry attempts)
+ * exception class name</li>
+ * </ul>
+ * <p>
+ * The {@link #setCustomTags(Map)} and {@link #setCustomTagsProvider(Function)} can be
+ * used to further customize tags on the timers.
+ *
+ * @author Artem Bilan
+ * @since 2.0.8
+ */
+public class MetricsRetryListener implements RetryListener {
+
+	public static final String TIMER_NAME = "spring.retry";
+
+	private final MeterRegistry meterRegistry;
+
+	private final Map<RetryCallback<?, ?>, Timer.Sample> retryContextToSample = new HashMap<>();
+
+	private final Map<String, String> customTags = new HashMap<>();
+
+	private final Meter.MeterProvider<Timer> retryMeterProvider;
+
+	private Function<RetryContext, Iterable<Tag>> customTagsProvider = retryContext -> Tags.empty();
+
+	/**
+	 * Construct an instance based on the provided {@link MeterRegistry}.
+	 * @param meterRegistry the {@link MeterRegistry} to use for timers.
+	 */
+	public MetricsRetryListener(MeterRegistry meterRegistry) {
+		Assert.notNull(meterRegistry, "'meterRegistry' must not be null");
+		this.meterRegistry = meterRegistry;
+		this.retryMeterProvider = Timer.builder(TIMER_NAME)
+			.description("Metrics for Spring RetryTemplate")
+			.withRegistry(this.meterRegistry);
+	}
+
+	/**
+	 * Supply a map of tags which are going to be used for all the timers managed by this
+	 * listener.
+	 * @param customTags the map of additional tags for all the timers.
+	 */
+	public void setCustomTags(Map<String, String> customTags) {
+		Assert.notNull(customTags, "'customTags' must not be null");
+		this.customTags.putAll(customTags);
+	}
+
+	/**
+	 * Supply a {@link Function} to build additional tags for all the timers based on the
+	 * {@link RetryContext}.
+	 * @param customTagsProvider the {@link Function} for additional tags with a
+	 * {@link RetryContext} scope.
+	 */
+	public void setCustomTagsProvider(Function<RetryContext, Iterable<Tag>> customTagsProvider) {
+		Assert.notNull(customTagsProvider, "'customTagsProvider' must not be null");
+		this.customTagsProvider = customTagsProvider;
+	}
+
+	@Override
+	public <T, E extends Throwable> boolean open(RetryContext context, RetryCallback<T, E> callback) {
+		this.retryContextToSample.put(callback, Timer.start(this.meterRegistry));
+		return true;
+	}
+
+	@Override
+	public <T, E extends Throwable> void close(RetryContext context, RetryCallback<T, E> callback,
+			Throwable throwable) {
+
+		Timer.Sample sample = this.retryContextToSample.remove(callback);
+
+		Assert.state(sample != null,
+				() -> String.format("No 'Timer.Sample' registered for '%s'. Was the 'open()' called?", context));
+
+		Tags retryTags = Tags.of("name", callback.getLabel())
+			.and("retry.count", "" + context.getRetryCount())
+			.and(this.customTags.entrySet().stream().map((entry) -> Tag.of(entry.getKey(), entry.getValue())).toList())
```

Comment:
If you use `Tags` instead of `Map` for `customTags` above, you can do this:
```suggestion
			.and(this.customTags)
```